src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -0,0 +1,265 @@
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace System.Threading.RateLimiting
+{
+    /// <summary>
+    /// <see cref="RateLimiter"/> implementation that helps manage concurrent access to a resource.
+    /// </summary>
+    public sealed class ConcurrencyLimiter : RateLimiter
+    {
+        private int _permitCount;
+        private int _queueCount;
+
+        private readonly ConcurrencyLimiterOptions _options;
+        private readonly Deque<RequestRegistration> _queue = new Deque<RequestRegistration>();
+
+        private static readonly ConcurrencyLease SuccessfulLease = new ConcurrencyLease(true, null, 0);
+        private static readonly ConcurrencyLease FailedLease = new ConcurrencyLease(false, null, 0);
+        private static readonly ConcurrencyLease QueueLimitLease = new ConcurrencyLease(false, null, 0, "Queue limit reached");
+
+        // Use the queue as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _queue;
+
+        /// <summary>
+        /// Initializes the <see cref="ConcurrencyLimiter"/>.
+        /// </summary>
+        /// <param name="options">Options to specify the behavior of the <see cref="ConcurrencyLimiter"/>.</param>
+        public ConcurrencyLimiter(ConcurrencyLimiterOptions options)
+        {
+            _options = options;
+            _permitCount = _options.PermitLimit;
+        }
+
+        /// <inheritdoc/>
+        public override int GetAvailablePermits() => _permitCount;
+
+        /// <inheritdoc/>
+        protected override RateLimitLease AcquireCore(int permitCount)
+        {
+            // These amounts of resources can never be acquired
+            if (permitCount > _options.PermitLimit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(permitCount), $"{permitCount} permits exceeds the permit limit of {_options.PermitLimit}.");
+            }
+
+            // Return SuccessfulLease or FailedLease to indicate limiter state
+            if (permitCount == 0)
+            {
+                return _permitCount > 0 ? SuccessfulLease : FailedLease;
+            }
+
+            // Perf: Check SemaphoreSlim implementation instead of locking
+            if (_permitCount >= permitCount)
+            {
+                lock (Lock)
+                {
+                    if (TryLeaseUnsynchronized(permitCount, out RateLimitLease? lease))
+                    {
+                        return lease;
+                    }
+                }
+            }
+
+            return FailedLease;
+        }
+
+        /// <inheritdoc/>
+        protected override ValueTask<RateLimitLease> WaitAsyncCore(int permitCount, CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Generally the exception that's emerging from here and all of the throws below should instead be stored in the returned task.

Reply:
That's to avoid throwing if the calling code stores the task and instead throw at the await site?

```csharp
// this can throw with current code
var task = limiter.WaitAsync();
```

Reply:
Yes, or pretty much does anything at the call site other than directly await'ing, e.g.
```C#
limiter.WaitAsync().ContinueWith(...);
```
or
```C#
Task.WaitAll(limiter1.WaitAsync(), limiter2.WaitAsync());
```
etc.

Reply:
Even the argument exceptions?

Reply:
Generally not for argument exceptions, if they're really that, e.g. if they represent a programmer/user error that's deterministic.  Those should typically come first, as well.